# Who Listens to What? An Exploratory Analysis of Music Preferences

In the world of music streaming, data tells a story. This project dives into a dataset of music listening habits from two cities, Springfield and Shelbyville, to uncover the hidden narratives within their playlists. The primary goal is to perform a comprehensive analysis to understand, compare, and contrast the musical preferences of their residents.

We will answer key questions such as:

- What are the most popular genres in each city?

- Do listening habits change on different days of the week?

- How do the overall music tastes of the two cities truly differ?

This notebook will walk through the entire data analysis process, starting with crucial data cleaning and preprocessing, moving into exploratory data analysis (EDA) with visualizations, and finally, performing hypothesis testing to validate our observations.

**Tools Used**: Python, Pandas, Matplotlib, and Seaborn.

## Step 1: Initial Data Inspection

Before we do anything else, it's important to load and inspect the data to get a clear picture of what we're working with. This helps us confirm column names, find potential issues, and understand the data types.

In [1]:
# Load and inspect the dataset
import pandas as pd

# Load the CSV file
df = pd.read_csv('data/music_project_en (1).csv')

# Display the first few rows
df.head()

userID                        Track            artist  genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile   rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg   rock   
2    20EC38            Funiculì funiculà       Mario Lanza    pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice   folk   
4  E2DC1FAE                  Soul People        Space Echo  dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday

In [2]:
# Check for leading/trailing spaces in column names
print('Column names:', list(df.columns))

# Check for missing values in each column
print('\nMissing values per column:')
print(df[['Track', 'artist', 'genre']].isnull().sum())

# Check for duplicate rows
num_duplicates = df.duplicated().sum()
print(f'\nNumber of duplicate rows: {num_duplicates}')

# Check data types and non-null counts
print('\nDataFrame info:')
df.info()

Column names: ['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day']

Missing values per column:
Track     1343
artist    7567
genre     1198
dtype: int64

Number of duplicate rows: 3826

DataFrame info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


### Data Inspection Findings
The initial look at the data gives us critical information:

- **Column Names:** There are leading/trailing spaces in some column names (e.g., `'  userID'`, `'  City  '`). This is a common issue and needs to be fixed to avoid errors.
- **Missing Values:** We have missing values in the `Track`, `artist`, and `genre` columns. We need a strategy to handle these. Simply dropping them might remove too much data, so we'll have to be smart about it.
- **Duplicates:** There are 3,826 duplicate rows. These should be removed to ensure our analysis is accurate.
- **Data Types:** All columns are currently of the object type. While this is fine for most, the `time` column should ideally be a datetime object if we want to do time-based analysis.